manager/controlapi/node.go

```diff
@@ -220,14 +240,19 @@ func (s *Server) UpdateNode(ctx context.Context, request *api.UpdateNodeRequest)
 	}
 
 	if demote && s.raft != nil {
-		memberlist := s.raft.GetMemberlist()
-		for raftID, member := range memberlist {
-			if member.NodeID == request.NodeID {
-				if err := s.raft.RemoveMember(ctx, raftID); err != nil {
-					return nil, err
+		if err := s.raft.RemoveMember(ctx, member.RaftID); err != nil {
+			// Rollback to the initial Spec if we can't finalize the role
+			// change by removing the member from raft.
+			s.store.Update(func(tx store.Tx) error {
+				node = store.GetNode(tx, request.NodeID)
+				if node == nil {
+					return nil
 				}
-				break
-			}
+
+				node.Spec = initSpec
```

Comment:
> The client may have changed other things in the spec. Here we're discarding all those other changes because the demotion failed.

I've asked for that in a previous comment, but imho I think we should not allow the Spec change if the role change fails. This gives a sense that the Spec change is `incomplete`, which can be confusing to the user that might expect the demotion to take effect as much as all the other changes from the Spec. And the issue here is that we don't try (yet) to reconcile the raft memberlist from what is in the Spec.

> The corner case @stevvooe points out here with node failure.

The corner case is definitely possible, but it's better than the current behavior which does not fix the role on `Remove` failure and leaves the node stuck to role changes.

I'm not against reverting to the either doing the rollback for the role only (Spec being partially applied), or removing this entirely. Even though we can leave this here as a best effort mechanism to rollback in case of a failure on `Removal`.

The real solution would be to make the demotion happen in the Spec and then try to `Remove` the node from raft until it succeeds to match the behavior of other Spec changes, but outside of the call to `UpdateNode`.